setup.py

```diff
@@ -1,73 +1,110 @@
+        # Since we force CMake via PEP 518 in the pyproject.toml, this should
+        # never happen and this whole method can be removed in your code if you
+        # want.
         try:
-            out = subprocess.check_output(['cmake', '--version'])
+            subprocess.check_output(["cmake", "--version"])
         except OSError:
-            raise RuntimeError("CMake must be installed to build the following extensions: " +
-                               ", ".join(e.name for e in self.extensions))
-
-        if platform.system() == "Windows":
-            cmake_version = LooseVersion(re.search(r'version\s*([\d.]+)', out.decode()).group(1))
-            if cmake_version < '3.1.0':
-                raise RuntimeError("CMake >= 3.1.0 is required on Windows")
+            msg = "CMake missing - probably upgrade to a newer version of Pip?"
+            raise RuntimeError(msg)
 
-        for ext in self.extensions:
-            self.build_extension(ext)
+        # To support Python 2, we have to avoid super(), since distutils is all
+        # old-style classes.
+        build_ext.run(self)
 
     def build_extension(self, ext):
         extdir = os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.name)))
+
         # required for auto-detection of auxiliary "native" libs
         if not extdir.endswith(os.path.sep):
             extdir += os.path.sep
 
-        cmake_args = ['-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=' + extdir,
-                      '-DPYTHON_EXECUTABLE=' + sys.executable]
+        cfg = "Debug" if self.debug else "Release"
 
-        cfg = 'Debug' if self.debug else 'Release'
-        build_args = ['--config', cfg]
+        # Set Python_EXECUTABLE instead if you use PYBIND11_FINDPYTHON
+        cmake_args = [
+            "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY={}".format(extdir),
+            "-DPYTHON_EXECUTABLE={}".format(sys.executable),
+            "-DEXAMPLE_VERSION_INFO={}".format(self.distribution.get_version()),
+            "-DCMAKE_BUILD_TYPE={}".format(cfg),  # not used on MSVC, but no harm
+        ]
+        build_args = []
+
+        if self.compiler.compiler_type != "msvc":
+            # Using Ninja-build since it a) is available as a wheel and b)
+            # multithreads automatically. MSVC would require all variables be
+            # exported for Ninja to pick it up, which is a little tricky to do.
+            cmake_args += ["-GNinja"]
 
-        if platform.system() == "Windows":
-            cmake_args += ['-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_{}={}'.format(cfg.upper(), extdir)]
-            if sys.maxsize > 2**32:
-                cmake_args += ['-A', 'x64']
-            build_args += ['--', '/m']
         else:
-            cmake_args += ['-DCMAKE_BUILD_TYPE=' + cfg]
-            build_args += ['--', '-j2']
+            # CMake lets you override the generator - we need to check this.
+            # Can be set with Conda-Build, for example.
```

Comment:
Is there a reason we only allow this for MSVC, and not for the others?

Reply:
Because we force Ninja otherwise (it gets downloaded in pyproject.toml), and the arguments are basically the same for all the non-multi-config generators anyway. I couldn't get Ninja to work with MSVC yet, so it's special.

Reply:
In my version of this, I pass the command line option `-j N` to setup.py to cmake. If no option is given, I use `multiprocessing.cpu_count()`. As mentioned before, I think ninja should not be required, since it is not essential.

Reply:
Ninja comes in wheels, just like CMake (Kitware makes both wheels, actually). You can't be sure that the computer running this will have make (for example, a docker image often does not). Ninja is the only generator you can ensure is always present, because you include it in your pyproject.toml along with cmake.

Reply:
Also, by the way, you can select other generators via environment variables (see #25), so unless you force the generator like I'm doing on non-Windows, then it's invalid to pass options directly through to the generator (via `--`); you should use CMake 3.14+'s -j option directly or verify that `CMAKE_GENERATOR` is not set. I think conda may set it, at least on Windows. And on my computer I always have `CMAKE_GENERATOR` set to `"Ninja"` since it's faster and always multithreads (I used to set MAKEFLAGS to `-j8` or however many cores I had instead).

TODO: I should only add `-GNinja` if `CMAKE_GENERATOR` is not set to allow this to be overridden on all platforms, possibly?

Reply:
FYI,  `multiprocessing.cpu_count()` may raise `NotImplementedError`. See https://docs.python.org/3.5/library/multiprocessing.html#multiprocessing.cpu_count . For iMinuit, `os.cpu_count()` (new in 3.4) will return None instead of raising if undetermined.